### ISSUE TITLE
Misleading diagnostics with missing bounds when deriving a trait 

### ISSUE THREAD

**Issue by a1phyr** [rust-lang/rust#84003]
Given the following code: https://play.rust-lang.org/?edition=2018&gist=5f8c9b8b72174c48746e236c827ab881

```rust
use std::fmt;

trait Trait {}

struct X<T>(T);

impl<T: fmt::Debug + Trait> fmt::Debug for X<T> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.debug_tuple("X").field(&self.0).finish()
    }
}

#[derive(Debug)]
struct Y<T>(X<T>);
```

The current output is:

```
error[E0277]: the trait bound `T: Trait` is not satisfied
  --> src/lib.rs:14:13
   |
14 | struct Y<T>(X<T>);
   |             ^^^^ the trait `Trait` is not implemented for `T`
   |
   = note: required because of the requirements on the impl of `Debug` for `X<T>`
   = note: 1 redundant requirements hidden
   = note: required because of the requirements on the impl of `Debug` for `&X<T>`
   = note: required for the cast to the object type `dyn Debug`
   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider further restricting this bound
   |
13 | #[derive(Debug + Trait)]
   |                ^^^^^^^
```

The proposed solution is obviously wrong, the compiler should suggest a manual impl instead.

**Comment by lo48576** [comment]
I found another variation, where adding `where` clause is suggested inside `derive`.

Source (<https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=20c02bc28c88fbc15614319fa7509fbd>):
```rust
use std::fmt;

trait DebugExt: fmt::Debug {}

pub struct Inner<T>(T);

impl fmt::Debug for Inner<()> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        todo!()
    }
}

#[derive(Debug)]
pub struct Outer<T>(Inner<T>);
```

Compiler output:
```
error[E0277]: `Inner<T>` doesn't implement `Debug`
  --> src/lib.rs:14:21
   |
13 | #[derive(Debug)]
   |          ----- in this derive macro expansion
14 | pub struct Outer<T>(Inner<T>);
   |                     ^^^^^^^^ `Inner<T>` cannot be formatted using `{:?}`
   |
   = help: the trait `Debug` is not implemented for `Inner<T>`
   = note: add `#[derive(Debug)]` to `Inner<T>` or manually `impl Debug for Inner<T>`
   = note: this error originates in the derive macro `Debug` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider introducing a `where` bound, but there might be an alternative better way to express this requirement
   |
13 | #[derive(Debug where Inner<T>: Debug)]
   |                +++++++++++++++++++++

For more information about this error, try `rustc --explain E0277`.
```
